File: apps/patter-voice-sdk/libraries/python/getpatter/observability/tracing.py
"""OpenTelemetry tracing helpers for Patter.

Design goals:
* **Zero cost when disabled.** ``start_span`` falls back to a cheap no-op
  context manager unless the opt-in env var ``PATTER_OTEL_ENABLED=1`` is set
  *and* the OTel SDK could be imported.
* **Opt-in only.** No telemetry is emitted by default. We never export PII
  (user utterances, tool payloads) as span attributes — only sizes and
  provider identifiers.
* **Single source of truth for span names.** Downstream services should
  attribute spans by the constants at the bottom of this module.

Spans used by Patter:

* ``getpatter.call`` — the top-level call span (created in the stream handler)
* ``getpatter.stt`` — a streamed STT inference
* ``getpatter.llm`` — an LLM completion (per turn)
* ``getpatter.tts`` — a TTS synthesis
* ``getpatter.tool`` — a single tool invocation
* ``getpatter.endpoint`` — silence-detected → LLM-dispatch window
* ``getpatter.bargein`` — interrupt-detected → TTS-stopped window
"""

from __future__ import annotations

import logging
import os
import sys
from contextlib import contextmanager
from typing import Any, Iterator

logger = logging.getLogger("getpatter.observability")

ENV_FLAG = "PATTER_OTEL_ENABLED"
SERVICE_NAME = "patter"

# --- Span names -------------------------------------------------------------
SPAN_CALL = "getpatter.call"
SPAN_STT = "getpatter.stt"
SPAN_LLM = "getpatter.llm"
SPAN_TTS = "getpatter.tts"
SPAN_TOOL = "getpatter.tool"
SPAN_ENDPOINT = "getpatter.endpoint"
SPAN_BARGEIN = "getpatter.bargein"

# --- State ------------------------------------------------------------------
_tracer: Any = None
_provider: Any = None
_initialized = False


def is_enabled() -> bool:
    """True only if the env flag is set *and* the tracer initialized cleanly."""
    return bool(_tracer) and os.getenv(ENV_FLAG, "").lower() in {"1", "true", "yes"}


def init_tracing(
    service_name: str = SERVICE_NAME,
    otlp_endpoint: str | None = None,
    resource_attributes: dict[str, str] | None = None,
) -> bool:
    """Install the OTel provider + OTLP exporter.

    Returns True if tracing was actually wired. Silently returns False if
    the env flag is off or the OTel packages are not installed.
    """
    global _tracer, _provider, _initialized

    if _initialized:
        return is_enabled()

    if os.getenv(ENV_FLAG, "").lower() not in {"1", "true", "yes"}:
        _initialized = True
        return False

    try:
        from opentelemetry import trace
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
    except ImportError:
        logger.warning(
            "PATTER_OTEL_ENABLED=1 but opentelemetry SDK not installed. "
            "Install with: pip install getpatter[tracing]"
        )
        _initialized = True
        return False

    attrs = {"service.name": service_name}
    if resource_attributes:
        attrs.update(resource_attributes)
    resource = Resource.create(attrs)
    _provider = TracerProvider(resource=resource)

    endpoint = otlp_endpoint or os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    try:
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
            OTLPSpanExporter,
        )

        exporter = (
            OTLPSpanExporter(endpoint=f"{endpoint}/v1/traces")
            if endpoint
            else OTLPSpanExporter()
        )
        _provider.add_span_processor(BatchSpanProcessor(exporter))
    except ImportError:
        logger.warning(
            "opentelemetry OTLP exporter not installed; spans will not be exported"
        )

    trace.set_tracer_provider(_provider)
    _tracer = trace.get_tracer(service_name)
    _initialized = True
    logger.info(
        "Patter OTel tracing enabled (service=%s, endpoint=%s)", service_name, endpoint
    )
    return True


def get_tracer() -> Any:
    """Returns the current tracer or None."""
    return _tracer


@contextmanager
def start_span(
    name: str,
    attributes: dict[str, Any] | None = None,
) -> Iterator[Any]:
    """Start a span with optional attributes.

    No-op (yields None) when tracing is disabled. Callers should tolerate a
    None span — see :meth:`_noop_span_attr`.
    """
    if not is_enabled():
        yield None
        return

    span_cm = _tracer.start_as_current_span(name)
    span = span_cm.__enter__()
    exc_info: tuple[Any, Any, Any] = (None, None, None)
    try:
        if attributes:
            for k, v in attributes.items():
                try:
                    span.set_attribute(k, v)
                except Exception:  # pragma: no cover
                    pass
        yield span
    except Exception as exc:
        exc_info = sys.exc_info()
        try:
            span.record_exception(exc)
        except Exception:  # pragma: no cover
            pass
        raise
    finally:
        try:
            span_cm.__exit__(*exc_info)
        except Exception:  # pragma: no cover
            pass


def shutdown_tracing() -> None:
    """Flush any pending spans. Safe to call unconditionally."""
    global _tracer, _provider, _initialized
    if _provider is not None:
        try:
            _provider.shutdown()
        except Exception:  # pragma: no cover
            pass
    _tracer = None
    _provider = None
    _initialized = False

File: apps/patter-voice-sdk/libraries/typescript/src/observability/tracing.ts
/**
 * OpenTelemetry tracing helpers for Patter — TypeScript mirror of
 * ``libraries/python/getpatter/observability/tracing.py``.
 *
 * Design goals:
 *   - Zero cost when disabled. ``startSpan`` falls back to a cheap no-op span
 *     unless the opt-in env var ``PATTER_OTEL_ENABLED=1`` is set *and* the
 *     optional ``@opentelemetry/api`` package is installed.
 *   - Opt-in only. No telemetry is emitted by default. We never export PII
 *     (user utterances, tool payloads) as span attributes — only sizes and
 *     provider identifiers.
 *   - Single source of truth for span names — downstream services should
 *     attribute spans using the ``SPAN_*`` constants at the bottom.
 */
import { getLogger } from '../logger';

/** Environment variable that gates all OpenTelemetry wire-up. */
export const ENV_FLAG = 'PATTER_OTEL_ENABLED';
/** Default `service.name` reported on exported spans. */
export const SERVICE_NAME = 'patter';

// --- Span names -------------------------------------------------------------
// Normalized to ``getpatter.*`` for parity with the Python SDK and to avoid
// the historical ``patter.*`` / ``getpatter.*`` mix that fragmented dashboards.
export const SPAN_CALL = 'getpatter.call';
export const SPAN_STT = 'getpatter.stt';
export const SPAN_LLM = 'getpatter.llm';
export const SPAN_TTS = 'getpatter.tts';
export const SPAN_TOOL = 'getpatter.tool';
export const SPAN_ENDPOINT = 'getpatter.endpoint';
export const SPAN_BARGEIN = 'getpatter.bargein';

/**
 * Minimal span surface area — subset of the OTel ``Span`` API the Patter SDK
 * relies on. We keep this narrow so the no-op fallback stays trivial.
 */
export interface Span {
  setAttribute(key: string, value: unknown): void;
  recordException(exception: unknown): void;
  end(): void;
}

/** Options for `initTracing()`. */
export interface InitTracingOptions {
  serviceName?: string;
  otlpEndpoint?: string;
  resourceAttributes?: Record<string, string>;
}

// --- OTel handle (lazily resolved) -----------------------------------------
interface OtelApiShape {
  trace: {
    getTracer(name: string): {
      startSpan(name: string, options?: { attributes?: Record<string, unknown> }): unknown;
    };
    setGlobalTracerProvider?(provider: unknown): void;
  };
}

/** Minimal surface of ``NodeTracerProvider`` / ``BasicTracerProvider``. */
interface TracerProviderShape {
  addSpanProcessor?(processor: unknown): void;
  register?(): void;
  shutdown?(): Promise<void>;
  forceFlush?(): Promise<void>;
}

let otel: OtelApiShape | null = null;
let initialized = false;
let tracerAvailable = false;
let provider: TracerProviderShape | null = null;

function tryLoadOtel(): OtelApiShape | null {
  if (otel !== null) return otel;
  try {
    // ``require`` keeps this an optional peer dep — bundlers that don't tree
    // shake this will still be fine because the module is only touched when
    // ``PATTER_OTEL_ENABLED`` is set.
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const mod = require('@opentelemetry/api') as OtelApiShape;
    otel = mod;
    return mod;
  } catch {
    return null;
  }
}

/**
 * Attempt to wire a ``NodeTracerProvider`` + OTLP HTTP exporter when the
 * optional SDK packages are installed. Returns ``null`` when any piece is
 * missing — callers fall back to the no-op tracer (current behaviour).
 */
function trySetupSdk(
  options: InitTracingOptions,
  api: OtelApiShape,
): TracerProviderShape | null {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const sdkTraceNode = require('@opentelemetry/sdk-trace-node') as {
      NodeTracerProvider: new (opts?: Record<string, unknown>) => TracerProviderShape;
    };
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const sdkTraceBase = require('@opentelemetry/sdk-trace-base') as {
      BatchSpanProcessor: new (exporter: unknown) => unknown;
    };
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const otlpHttp = require('@opentelemetry/exporter-trace-otlp-http') as {
      OTLPTraceExporter: new (opts?: { url?: string }) => unknown;
    };

    const serviceName = options.serviceName ?? SERVICE_NAME;
    // Many SDK versions accept a ``resource`` object, but building one pulls
    // in yet another package. Pass service name via options when the provider
    // supports it — otherwise rely on OTEL_SERVICE_NAME env.
    const providerInstance = new sdkTraceNode.NodeTracerProvider({
      resource: {
        attributes: {
          'service.name': serviceName,
          ...(options.resourceAttributes ?? {}),
        },
      },
    });

    const endpoint =
      options.otlpEndpoint ?? process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? undefined;
    const exporter = new otlpHttp.OTLPTraceExporter(
      endpoint ? { url: `${endpoint.replace(/\/$/, '')}/v1/traces` } : undefined,
    );
    const processor = new sdkTraceBase.BatchSpanProcessor(exporter);
    providerInstance.addSpanProcessor?.(processor);
    providerInstance.register?.();

    // Best-effort: expose globally so ``api.trace.getTracer`` returns this
    // provider's tracer.
    try {
      api.trace.setGlobalTracerProvider?.(providerInstance);
    } catch {
      // Swallow — ``register()`` already sets the global in most SDKs.
    }
    return providerInstance;
  } catch (e) {
    getLogger().debug(
      `[observability] OTel SDK wire-up skipped: ${String((e as Error)?.message ?? e)}`,
    );
    return null;
  }
}

function envFlagEnabled(): boolean {
  const raw = (process.env[ENV_FLAG] ?? '').toLowerCase();
  return raw === '1' || raw === 'true' || raw === 'yes';
}

/**
 * Initialize tracing. Returns ``true`` when OTel is wired, ``false`` otherwise
 * (which covers both "env flag off" and "peer dep missing").
 *
 * If the optional SDK packages (``@opentelemetry/sdk-trace-node``,
 * ``@opentelemetry/sdk-trace-base``, ``@opentelemetry/exporter-trace-otlp-http``)
 * are installed, a ``NodeTracerProvider`` with OTLP/HTTP exporter is wired up
 * automatically. Otherwise, spans produced via ``startSpan`` are still created
 * against whatever global provider ``@opentelemetry/api`` resolves to (which
 * may be a no-op if the host hasn't registered one).
 */
export function initTracing(options: InitTracingOptions = {}): boolean {
  if (initialized) return tracerAvailable;
  initialized = true;

  if (!envFlagEnabled()) {
    tracerAvailable = false;
    return false;
  }

  const api = tryLoadOtel();
  if (!api) {
    getLogger().warn(
      `${ENV_FLAG}=1 but @opentelemetry/api is not installed. ` +
        'Install with: npm install @opentelemetry/api ' +
        '@opentelemetry/sdk-trace-node @opentelemetry/sdk-trace-base ' +
        '@opentelemetry/exporter-trace-otlp-http',
    );
    tracerAvailable = false;
    return false;
  }

  // Optional SDK wire-up — fail silently to a no-op tracer when not installed.
  provider = trySetupSdk(options, api);

  tracerAvailable = true;
  const serviceName = options.serviceName ?? SERVICE_NAME;
  getLogger().info(
    `[observability] Patter OTel tracing enabled (service=${serviceName}${
      provider ? ', exporter=otlp-http' : ', exporter=noop'
    })`,
  );
  return true;
}

/**
 * Flush any pending spans and tear down the tracer provider. Safe to call
 * unconditionally — returns immediately when tracing was never wired up.
 */
export async function shutdownTracing(): Promise<void> {
  if (provider) {
    try {
      await provider.forceFlush?.();
    } catch {
      // Swallow.
    }
    try {
      await provider.shutdown?.();
    } catch {
      // Swallow.
    }
  }
  provider = null;
  otel = null;
  initialized = false;
  tracerAvailable = false;
}

/**
 * Return the configured tracer (or a cheap no-op when tracing is disabled).
 * The returned object exposes only ``startSpan`` — callers that need richer
 * OTel APIs should import ``@opentelemetry/api`` directly.
 */
export function getTracer(name: string = SERVICE_NAME): {
  startSpan(n: string, attrs?: Record<string, unknown>): Span;
} {
  if (!isTracingEnabled() || !otel) {
    return {
      startSpan: () => NOOP_SPAN,
    };
  }
  try {
    const tracer = otel.trace.getTracer(name);
    return {
      startSpan: (spanName: string, attrs?: Record<string, unknown>) => {
        try {
          const raw = tracer.startSpan(
            spanName,
            attrs ? { attributes: attrs } : undefined,
          );
          return new RealSpan(raw as OtelSpan);
        } catch {
          return NOOP_SPAN;
        }
      },
    };
  } catch {
    return { startSpan: () => NOOP_SPAN };
  }
}

/** True only if the env flag is set AND the tracer initialized cleanly. */
export function isTracingEnabled(): boolean {
  return tracerAvailable && envFlagEnabled();
}

// --- Noop fallback ----------------------------------------------------------
class NoopSpan implements Span {
  setAttribute(_key: string, _value: unknown): void {
    // no-op
  }
  recordException(_exception: unknown): void {
    // no-op
  }
  end(): void {
    // no-op
  }
}

const NOOP_SPAN = new NoopSpan();

// --- Real span wrapper ------------------------------------------------------
interface OtelSpan {
  setAttribute(key: string, value: unknown): unknown;
  recordException(exception: unknown): unknown;
  end(): unknown;
}

class RealSpan implements Span {
  private readonly span: OtelSpan;

  constructor(span: OtelSpan) {
    this.span = span;
  }

  setAttribute(key: string, value: unknown): void {
    try {
      this.span.setAttribute(key, value);
    } catch {
      // Swallow — OTel should never crash the call path.
    }
  }

  recordException(exception: unknown): void {
    try {
      this.span.recordException(exception);
    } catch {
      // Swallow.
    }
  }

  end(): void {
    try {
      this.span.end();
    } catch {
      // Swallow.
    }
  }
}

/**
 * Start a span. Callers must ``end()`` the returned span — use try/finally:
 *
 * ```ts
 * const span = startSpan(SPAN_LLM, { 'llm.model': 'gpt-4o' });
 * try { ... } finally { span.end(); }
 * ```
 *
 * Returns a no-op span when tracing is disabled or unavailable.
 */
export function startSpan(
  name: string,
  attrs?: Record<string, unknown>,
): Span {
  if (!isTracingEnabled() || !otel) return NOOP_SPAN;
  try {
    const tracer = otel.trace.getTracer(SERVICE_NAME);
    const rawSpan = tracer.startSpan(name, attrs ? { attributes: attrs } : undefined);
    return new RealSpan(rawSpan as OtelSpan);
  } catch {
    return NOOP_SPAN;
  }
}

/**
 * Convenience wrapper — starts a span, runs ``fn``, records exceptions on
 * throw, and always ends the span (try/finally). Mirrors Python's
 * ``with start_span(...):`` context-manager ergonomics.
 *
 * ```ts
 * await withSpan(SPAN_LLM, { 'llm.model': 'gpt-4o' }, async (span) => {
 *   span.setAttribute('llm.tokens', 123);
 *   return await callLLM();
 * });
 * ```
 */
export async function withSpan<T>(
  name: string,
  attrs: Record<string, unknown> | undefined,
  fn: (span: Span) => Promise<T>,
): Promise<T> {
  const span = startSpan(name, attrs);
  try {
    return await fn(span);
  } catch (exc) {
    span.recordException(exc);
    throw exc;
  } finally {
    span.end();
  }
}

/** Internal: reset module state (primarily for tests; not part of the public API). */
export function _resetTracingForTesting(): void {
  otel = null;
  initialized = false;
  tracerAvailable = false;
  provider = null;
}
